File: src/qnx/battery_ai_predictor.cpp
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <thread>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// Battery performance data structure
struct BatteryPerformance {
    float timestamp;
    float pack_voltage;
    float pack_current;
    float cell_temp;
    float capacity_remaining;  // Estimated capacity remaining (%)
    int cycle_count;          // Battery cycle count
    float age_months;         // Battery age in months
    float health_score;       // Overall battery health (0-100)
    
    BatteryPerformance() : timestamp(0), pack_voltage(0), pack_current(0), 
                          cell_temp(0), capacity_remaining(100), cycle_count(0), 
                          age_months(0), health_score(100) {}
};

// AI prediction results
struct BatteryPrediction {
    float remaining_life_hours;     // Predicted remaining life in hours
    float remaining_cycles;         // Predicted remaining cycles
    float degradation_rate;         // Rate of capacity loss per cycle
    
    BatteryPrediction() : remaining_life_hours(0), remaining_cycles(0), 
                         degradation_rate(0) {}
};

class BatteryAIPredictor {
private:
    std::string gemini_api_key;
    std::string gemini_api_url;
    std::vector<BatteryPerformance> performance_history;
    std::vector<BatteryPrediction> prediction_history;
    std::ofstream log_file;
    bool ai_enabled;
    
    // Battery specifications
    struct BatterySpecs {
        float nominal_capacity_ah;
        float nominal_voltage;
        int max_cycles;
        float max_temp;
        float min_temp;
        float max_current;
        
        BatterySpecs() : nominal_capacity_ah(100), nominal_voltage(355.2), 
                        max_cycles(1000), max_temp(60), min_temp(-20), max_current(500) {}
    } specs;

public:
    BatteryAIPredictor(const std::string& api_key = "") 
        : gemini_api_key(api_key), ai_enabled(!api_key.empty()) {
        gemini_api_url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent";
        log_file.open("battery_ai_predictor.log", std::ios::app);
        
        // Load battery specifications
        load_battery_specs();
    }
    
    ~BatteryAIPredictor() {
        if (log_file.is_open()) {
            log_file.close();
        }
    }
    
    void add_performance_data(const BatteryPerformance& data) {
        performance_history.push_back(data);
        
        // Keep only last 1000 entries to manage memory
        if (performance_history.size() > 1000) {
            performance_history.erase(performance_history.begin());
        }
        
        log_info("Added performance data: V=" + std::to_string(data.pack_voltage) + 
                "V, I=" + std::to_string(data.pack_current) + "A, Health=" + 
                std::to_string(data.health_score) + "%");
    }
    
    BatteryPrediction predict_battery_life() {
        if (performance_history.size() < 10) {
            log_warning("Insufficient data for prediction. Need at least 10 data points.");
            return BatteryPrediction();
        }
        
        if (!ai_enabled) {
            return predict_battery_life_analytical();
        }
        
        return predict_battery_life_ai();
    }
    
    void update_battery_specs(const BatterySpecs& new_specs) {
        specs = new_specs;
        log_info("Battery specifications updated");
    }
    
    void enable_ai(const std::string& api_key) {
        gemini_api_key = api_key;
        ai_enabled = true;
        log_info("AI prediction enabled");
    }
    
    void disable_ai() {
        ai_enabled = false;
        log_info("AI prediction disabled");
    }
    
    void print_prediction_report() {
        if (prediction_history.empty()) {
            std::cout << "No predictions available." << std::endl;
            return;
        }
        
        const auto& latest = prediction_history.back();
        
        std::cout << "\n=== AI Battery Life Prediction Report ===" << std::endl;
        std::cout << "Remaining Life: " << std::fixed << std::setprecision(1) 
                  << latest.remaining_life_hours << " hours" << std::endl;
        std::cout << "Remaining Cycles: " << std::fixed << std::setprecision(0) 
                  << latest.remaining_cycles << " cycles" << std::endl;
        std::cout << "Degradation Rate: " << std::fixed << std::setprecision(3) 
                  << latest.degradation_rate << "% per cycle" << std::endl;
        std::cout << "==========================================" << std::endl;
    }

private:
    BatteryPrediction predict_battery_life_ai() {
        BatteryPrediction prediction;
        
        try {
            // Prepare data for AI analysis
            std::string analysis_data = prepare_ai_analysis_data();
            
            // Create prompt for Gemini
            std::string prompt = create_ai_prompt(analysis_data);
            
            // Call Gemini API
            std::string ai_response = call_gemini_api(prompt);
            
            // Parse AI response
            prediction = parse_ai_response(ai_response);
            
        } catch (const std::exception& e) {
            log_error("AI prediction failed: " + std::string(e.what()));
            return predict_battery_life_analytical();
        }
        
        prediction_history.push_back(prediction);
        return prediction;
    }
    
    BatteryPrediction predict_battery_life_analytical() {
        BatteryPrediction prediction;
        
        if (performance_history.size() < 5) {
            return prediction;
        }
        
        // Calculate degradation rate
        float avg_health = 0;
        float avg_voltage = 0;
        float avg_temp = 0;
        
        for (const auto& data : performance_history) {
            avg_health += data.health_score;
            avg_voltage += data.pack_voltage;
            avg_temp += data.cell_temp;
        }
        
        avg_health /= performance_history.size();
        avg_voltage /= performance_history.size();
        avg_temp /= performance_history.size();
        
        // Estimate degradation rate based on health and temperature
        prediction.degradation_rate = calculate_degradation_rate(avg_health, avg_temp);
        
        // Calculate remaining cycles
        float current_cycles = performance_history.back().cycle_count;
        prediction.remaining_cycles = std::max(0.0f, specs.max_cycles - current_cycles);
        
        // Calculate remaining life based on current usage pattern
        prediction.remaining_life_hours = calculate_remaining_life_hours();
        
        prediction_history.push_back(prediction);
        return prediction;
    }
    
    std::string prepare_ai_analysis_data() {
        std::ostringstream data;
        data << "Battery Performance Analysis Data:\n\n";
        
        // Recent performance data
        data << "Recent Performance (last 10 readings):\n";
        int start = std::max(0, (int)performance_history.size() - 10);
        for (int i = start; i < performance_history.size(); i++) {
            const auto& perf = performance_history[i];
            data << "Time: " << perf.timestamp << "s, "
                 << "Voltage: " << perf.pack_voltage << "V, "
                 << "Current: " << perf.pack_current << "A, "
                 << "Temp: " << perf.cell_temp << "°C, "
                 << "Health: " << perf.health_score << "%, "
                 << "Cycles: " << perf.cycle_count << "\n";
        }
        
        // Battery specifications
        data << "\nBattery Specifications:\n";
        data << "Nominal Capacity: " << specs.nominal_capacity_ah << " Ah\n";
        data << "Nominal Voltage: " << specs.nominal_voltage << " V\n";
        data << "Max Cycles: " << specs.max_cycles << "\n";
        data << "Max Temperature: " << specs.max_temp << "°C\n";
        data << "Min Temperature: " << specs.min_temp << "°C\n";
        data << "Max Current: " << specs.max_current << " A\n";
        
        // Performance statistics
        data << "\nPerformance Statistics:\n";
        data << "Total Data Points: " << performance_history.size() << "\n";
        data << "Average Health: " << calculate_average_health() << "%\n";
        data << "Average Temperature: " << calculate_average_temperature() << "°C\n";
        data << "Temperature Range: " << calculate_temperature_range() << "°C\n";
        
        return data.str();
    }
    
    std::string create_ai_prompt(const std::string& analysis_data) {
        std::ostringstream prompt;
        prompt << "You are an expert battery systems engineer. Analyze the following battery performance data and provide predictions for battery life and health.\n\n";
        prompt << analysis_data << "\n\n";
        prompt << "Please provide a detailed analysis including:\n";
        prompt << "1. Predicted remaining battery life in hours\n";
        prompt << "2. Estimated remaining charge cycles\n";
        prompt << "3. Current degradation rate\n";
        prompt << "Format your response as JSON with the following structure:\n";
        prompt << "{\n";
        prompt << "  \"remaining_life_hours\": <float>,\n";
        prompt << "  \"remaining_cycles\": <float>,\n";
        prompt << "  \"degradation_rate\": <float>,\n";
        prompt << "}\n\n";
        prompt << "Consider factors like temperature effects, cycle count, voltage patterns, and aging when making your predictions.";
        
        return prompt.str();
    }
    
    std::string call_gemini_api(const std::string& prompt) {
        if (gemini_api_key.empty()) {
            throw std::runtime_error("Gemini API key not provided");
        }
        
        CURL* curl = curl_easy_init();
        if (!curl) {
            throw std::runtime_error("Failed to initialize CURL");
        }
        
        // Prepare JSON request
        json request;
        request["contents"][0]["parts"][0]["text"] = prompt;
        
        std::string json_request = request.dump();
        
        // Set up CURL
        struct curl_slist* headers = nullptr;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        headers = curl_slist_append(headers, ("Authorization: Bearer " + gemini_api_key).c_str());
        
        std::string response;
        
        curl_easy_setopt(curl, CURLOPT_URL, gemini_api_url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_request.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
        
        CURLcode res = curl_easy_perform(curl);
        
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
        
        if (res != CURLE_OK) {
            throw std::runtime_error("CURL request failed: " + std::string(curl_easy_strerror(res)));
        }
        
        return response;
    }
    
    BatteryPrediction parse_ai_response(const std::string& response) {
        BatteryPrediction prediction;
        
        try {
            json response_json = json::parse(response);
            
            // Extract the text content from Gemini response
            std::string content = response_json["candidates"][0]["content"]["parts"][0]["text"];
            
            // Try to extract JSON from the response
            size_t json_start = content.find('{');
            size_t json_end = content.rfind('}');
            
            if (json_start != std::string::npos && json_end != std::string::npos) {
                std::string json_str = content.substr(json_start, json_end - json_start + 1);
                json prediction_json = json::parse(json_str);
                
                prediction.remaining_life_hours = prediction_json.value("remaining_life_hours", 0.0f);
                prediction.remaining_cycles = prediction_json.value("remaining_cycles", 0.0f);
                prediction.degradation_rate = prediction_json.value("degradation_rate", 0.0f);
            } else {
                // Fallback parsing if JSON not found
                log_warning("Could not parse JSON from AI response, using analytical fallback");
            }
            
        } catch (const std::exception& e) {
            log_error("Failed to parse AI response: " + std::string(e.what()));
        }
        
        return prediction;
    }
    
    float calculate_degradation_rate(float avg_health, float avg_temp) {
        // Base degradation rate
        float base_rate = (100.0f - avg_health) / 100.0f * 0.1f;
        
        // Temperature factor
        float temp_factor = 1.0f;
        if (avg_temp > 45.0f) {
            temp_factor = 1.5f + (avg_temp - 45.0f) * 0.1f;
        } else if (avg_temp < 10.0f) {
            temp_factor = 1.2f;
        }
        
        return base_rate * temp_factor;
    }
    
    float calculate_remaining_life_hours() {
        if (performance_history.size() < 2) return 0.0f;
        
        // Calculate average current draw
        float avg_current = 0;
        for (const auto& data : performance_history) {
            avg_current += std::abs(data.pack_current);
        }
        avg_current /= performance_history.size();
        
        if (avg_current <= 0) return 0.0f;
        
        // Estimate remaining capacity
        float remaining_capacity = performance_history.back().capacity_remaining / 100.0f * specs.nominal_capacity_ah;
        
        return remaining_capacity / avg_current;
    }
    

    

    
    float calculate_average_health() {
        if (performance_history.empty()) return 0.0f;
        
        float sum = 0;
        for (const auto& data : performance_history) {
            sum += data.health_score;
        }
        return sum / performance_history.size();
    }
    
    float calculate_average_temperature() {
        if (performance_history.empty()) return 0.0f;
        
        float sum = 0;
        for (const auto& data : performance_history) {
            sum += data.cell_temp;
        }
        return sum / performance_history.size();
    }
    
    float calculate_temperature_range() {
        if (performance_history.empty()) return 0.0f;
        
        float min_temp = performance_history[0].cell_temp;
        float max_temp = performance_history[0].cell_temp;
        
        for (const auto& data : performance_history) {
            min_temp = std::min(min_temp, data.cell_temp);
            max_temp = std::max(max_temp, data.cell_temp);
        }
        
        return max_temp - min_temp;
    }
    
    void load_battery_specs() {
        // Load from file if available, otherwise use defaults
        std::ifstream file("battery_specs.json");
        if (file.is_open()) {
            try {
                json specs_json;
                file >> specs_json;
                
                specs.nominal_capacity_ah = specs_json.value("nominal_capacity_ah", 100.0f);
                specs.nominal_voltage = specs_json.value("nominal_voltage", 355.2f);
                specs.max_cycles = specs_json.value("max_cycles", 1000);
                specs.max_temp = specs_json.value("max_temp", 60.0f);
                specs.min_temp = specs_json.value("min_temp", -20.0f);
                specs.max_current = specs_json.value("max_current", 500.0f);
                
                log_info("Battery specifications loaded from file");
            } catch (const std::exception& e) {
                log_error("Failed to load battery specs: " + std::string(e.what()));
            }
        }
    }
    
    static size_t write_callback(void* contents, size_t size, size_t nmemb, std::string* userp) {
        userp->append((char*)contents, size * nmemb);
        return size * nmemb;
    }
    
    void log_info(const std::string& message) {
        log_message("INFO", message);
    }
    
    void log_warning(const std::string& message) {
        log_message("WARNING", message);
    }
    
    void log_error(const std::string& message) {
        log_message("ERROR", message);
    }
    
    void log_message(const std::string& level, const std::string& message) {
        auto now = std::chrono::system_clock::now();
        auto time_t = std::chrono::system_clock::to_time_t(now);
        std::string timestamp = std::ctime(&time_t);
        timestamp.pop_back(); // Remove newline
        
        std::string log_entry = timestamp + " [" + level + "] " + message;
        
        if (log_file.is_open()) {
            log_file << log_entry << std::endl;
            log_file.flush();
        }
        
        std::cout << log_entry << std::endl;
    }
};

// Example usage and testing
int main() {
    std::cout << "=== Battery AI Predictor Test ===" << std::endl;
    
    // Initialize predictor (with or without API key)
    BatteryAIPredictor predictor(""); // Empty for analytical mode
    
    // Add some sample performance data
    for (int i = 0; i < 20; i++) {
        BatteryPerformance data;
        data.timestamp = i * 3600.0f; // 1 hour intervals
        data.pack_voltage = 350.0f + (i % 10) * 2.0f; // Varying voltage
        data.pack_current = 50.0f + (i % 5) * 10.0f; // Varying current
        data.cell_temp = 25.0f + (i % 3) * 5.0f; // Varying temperature
        data.capacity_remaining = 85.0f - i * 0.5f; // Decreasing capacity
        data.cycle_count = i * 10; // Increasing cycles
        data.age_months = 6.0f + i * 0.5f; // Aging battery
        data.health_score = 90.0f - i * 0.8f; // Declining health
        
        predictor.add_performance_data(data);
    }
    
    // Make prediction
    BatteryPrediction prediction = predictor.predict_battery_life();
    
    // Print results
    predictor.print_prediction_report();
    
    std::cout << "\nTest completed." << std::endl;
    return 0;
} 

File: client/api/battery_ai_predictor.py
import requests
import json
import logging
from datetime import datetime, timedelta
from typing import List, Dict, Optional
from dataclasses import dataclass
import os
from pymongo import MongoClient
import asyncio
from motor.motor_asyncio import AsyncIOMotorClient

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@dataclass
class BatteryPerformance:
    timestamp: float
    pack_voltage: float
    pack_current: float
    cell_temp: float
    capacity_remaining: float = 100.0
    cycle_count: int = 0
    age_months: float = 0.0
    health_score: float = 100.0

@dataclass
class BatteryPrediction:
    timestamp: float
    remaining_life_hours: float
    remaining_cycles: float
    degradation_rate: float
    confidence_score: float = 0.0
    recommendations: List[str] = None

@dataclass
class BatterySpecs:
    nominal_capacity_ah: float = 100.0
    nominal_voltage: float = 355.2
    max_cycles: int = 1000
    max_temp: float = 60.0
    min_temp: float = -20.0
    max_current: float = 500.0
    chemistry: str = "LiFePO4"
    cell_configuration: str = "96S72P"

class BatteryAIPredictor:
    def __init__(self, gemini_api_key: str = "", mongo_uri: str = ""):
        self.gemini_api_key = gemini_api_key or os.getenv("GEMINI_API_KEY", "")
        self.gemini_api_url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent"
        self.ai_enabled = bool(self.gemini_api_key)
        
        # MongoDB connection
        self.mongo_uri = mongo_uri or os.getenv("MONGODB_URI", "")
        self.db = None
        if self.mongo_uri:
            try:
                self.client = AsyncIOMotorClient(self.mongo_uri)
                self.db = self.client.battery_db
                logger.info("Connected to MongoDB for AI predictor")
            except Exception as e:
                logger.error(f"Failed to connect to MongoDB: {e}")
        
        # Battery specifications
        self.specs = self.load_battery_specs()
        
        # Performance history (in-memory cache)
        self.performance_history: List[BatteryPerformance] = []
        self.prediction_history: List[BatteryPrediction] = []
        
        logger.info(f"Battery AI Predictor initialized. AI enabled: {self.ai_enabled}")
    
    def load_battery_specs(self) -> BatterySpecs:
        """Load battery specifications from file or use defaults"""
        try:
            with open("battery_specs.json", "r") as f:
                specs_data = json.load(f)
                return BatterySpecs(**specs_data.get("battery_specifications", {}))
        except FileNotFoundError:
            logger.warning("battery_specs.json not found, using defaults")
            return BatterySpecs()
        except Exception as e:
            logger.error(f"Error loading battery specs: {e}")
            return BatterySpecs()
    
    async def add_performance_data(self, data: BatteryPerformance):
        """Add new performance data and update history"""
        self.performance_history.append(data)
        
        # Keep only last 1000 entries to manage memory
        if len(self.performance_history) > 1000:
            self.performance_history.pop(0)
        
        # Store in MongoDB if available
        if self.db:
            try:
                await self.db.battery_performance.insert_one({
                    "timestamp": data.timestamp,
                    "pack_voltage": data.pack_voltage,
                    "pack_current": data.pack_current,
                    "cell_temp": data.cell_temp,
                    "capacity_remaining": data.capacity_remaining,
                    "cycle_count": data.cycle_count,
                    "age_months": data.age_months,
                    "health_score": data.health_score,
                    "source": "battery_ai_predictor"
                })
            except Exception as e:
                logger.error(f"Failed to store performance data: {e}")
        
        logger.info(f"Added performance data: V={data.pack_voltage}V, "
                   f"I={data.pack_current}A, Health={data.health_score}%")
    
    async def predict_battery_life(self) -> BatteryPrediction:
        """Generate battery life prediction using AI or analytical methods"""
        if len(self.performance_history) < 10:
            logger.warning("Insufficient data for prediction. Need at least 10 data points.")
            return BatteryPrediction(
                timestamp=datetime.now().timestamp(),
                remaining_life_hours=0,
                remaining_cycles=0,
                degradation_rate=0,
                confidence_score=0
            )
        
        if self.ai_enabled:
            return await self.predict_battery_life_ai()
        else:
            return self.predict_battery_life_analytical()
    
    async def predict_battery_life_ai(self) -> BatteryPrediction:
        """Use Gemini AI for battery life prediction"""
        try:
            # Prepare data for AI analysis
            analysis_data = self.prepare_ai_analysis_data()
            
            # Create prompt for Gemini
            prompt = self.create_ai_prompt(analysis_data)
            
            # Call Gemini API
            ai_response = await self.call_gemini_api(prompt)
            
            # Parse AI response
            prediction = self.parse_ai_response(ai_response)
            
        except Exception as e:
            logger.error(f"AI prediction failed: {e}")
            prediction = self.predict_battery_life_analytical()
        
        # Store prediction
        self.prediction_history.append(prediction)
        
        # Store in MongoDB if available
        if self.db:
            try:
                await self.db.battery_predictions.insert_one({
                    "timestamp": prediction.timestamp,
                    "remaining_life_hours": prediction.remaining_life_hours,
                    "remaining_cycles": prediction.remaining_cycles,
                    "degradation_rate": prediction.degradation_rate,
                    "confidence_score": prediction.confidence_score,
                    "recommendations": prediction.recommendations or [],
                    "source": "battery_ai_predictor"
                })
            except Exception as e:
                logger.error(f"Failed to store prediction: {e}")
        
        return prediction
    
    def predict_battery_life_analytical(self) -> BatteryPrediction:
        """Use analytical methods for battery life prediction"""
        if len(self.performance_history) < 5:
            return BatteryPrediction(
                timestamp=datetime.now().timestamp(),
                remaining_life_hours=0,
                remaining_cycles=0,
                degradation_rate=0,
                confidence_score=0
            )
        
        # Calculate averages
        avg_health = sum(d.health_score for d in self.performance_history) / len(self.performance_history)
        avg_voltage = sum(d.pack_voltage for d in self.performance_history) / len(self.performance_history)
        avg_temp = sum(d.cell_temp for d in self.performance_history) / len(self.performance_history)
        
        # Estimate degradation rate
        degradation_rate = self.calculate_degradation_rate(avg_health, avg_temp)
        
        # Calculate remaining cycles
        current_cycles = self.performance_history[-1].cycle_count
        remaining_cycles = max(0, self.specs.max_cycles - current_cycles)
        
        # Calculate remaining life
        remaining_life_hours = self.calculate_remaining_life_hours()
        
        prediction = BatteryPrediction(
            timestamp=datetime.now().timestamp(),
            remaining_life_hours=remaining_life_hours,
            remaining_cycles=remaining_cycles,
            degradation_rate=degradation_rate,
            confidence_score=0.7,  # Medium confidence for analytical
            recommendations=self.generate_recommendations(avg_health, avg_temp)
        )
        
        self.prediction_history.append(prediction)
        return prediction
    
    def prepare_ai_analysis_data(self) -> str:
        """Prepare battery data for AI analysis"""
        data = []
        data.append("Battery Performance Analysis Data:\n\n")
        
        # Recent performance data
        data.append("Recent Performance (last 10 readings):\n")
        start = max(0, len(self.performance_history) - 10)
        for i in range(start, len(self.performance_history)):
            perf = self.performance_history[i]
            data.append(f"Time: {perf.timestamp}s, "
                       f"Voltage: {perf.pack_voltage}V, "
                       f"Current: {perf.pack_current}A, "
                       f"Temp: {perf.cell_temp}°C, "
                       f"Health: {perf.health_score}%, "
                       f"Cycles: {perf.cycle_count}\n")
        
        # Battery specifications
        data.append(f"\nBattery Specifications:\n")
        data.append(f"Nominal Capacity: {self.specs.nominal_capacity_ah} Ah\n")
        data.append(f"Nominal Voltage: {self.specs.nominal_voltage} V\n")
        data.append(f"Max Cycles: {self.specs.max_cycles}\n")
        data.append(f"Max Temperature: {self.specs.max_temp}°C\n")
        data.append(f"Min Temperature: {self.specs.min_temp}°C\n")
        data.append(f"Max Current: {self.specs.max_current} A\n")
        data.append(f"Chemistry: {self.specs.chemistry}\n")
        
        # Performance statistics
        data.append(f"\nPerformance Statistics:\n")
        data.append(f"Total Data Points: {len(self.performance_history)}\n")
        data.append(f"Average Health: {self.calculate_average_health()}%\n")
        data.append(f"Average Temperature: {self.calculate_average_temperature()}°C\n")
        data.append(f"Temperature Range: {self.calculate_temperature_range()}°C\n")
        
        return "".join(data)
    
    def create_ai_prompt(self, analysis_data: str) -> str:
        """Create prompt for Gemini AI"""
        prompt = f"""You are an expert battery systems engineer. Analyze the following battery performance data and provide predictions for battery life and health.

{analysis_data}

Please provide a detailed analysis including:
1. Predicted remaining battery life in hours
2. Estimated remaining charge cycles
3. Current degradation rate
4. Confidence score (0-1)
5. Specific recommendations for battery maintenance

Format your response as JSON with the following structure:
{{
  "remaining_life_hours": <float>,
  "remaining_cycles": <float>,
  "degradation_rate": <float>,
  "confidence_score": <float>,
  "recommendations": ["<string>", "<string>"]
}}

Consider factors like temperature effects, cycle count, voltage patterns, and aging when making your predictions."""
        
        return prompt
    
    async def call_gemini_api(self, prompt: str) -> str:
        """Call Gemini API with the given prompt"""
        if not self.gemini_api_key:
            raise ValueError("Gemini API key not provided")
        
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.gemini_api_key}"
        }
        
        payload = {
            "contents": [{
                "parts": [{"text": prompt}]
            }]
        }
        
        try:
            response = requests.post(
                self.gemini_api_url,
                headers=headers,
                json=payload,
                timeout=30
            )
            response.raise_for_status()
            return response.text
        except requests.exceptions.RequestException as e:
            raise Exception(f"API request failed: {e}")
    
    def parse_ai_response(self, response: str) -> BatteryPrediction:
        """Parse AI response and extract prediction data"""
        try:
            response_data = json.loads(response)
            content = response_data["candidates"][0]["content"]["parts"][0]["text"]
            
            # Try to extract JSON from the response
            json_start = content.find('{')
            json_end = content.rfind('}')
            
            if json_start != -1 and json_end != -1:
                json_str = content[json_start:json_end + 1]
                prediction_data = json.loads(json_str)
                
                return BatteryPrediction(
                    timestamp=datetime.now().timestamp(),
                    remaining_life_hours=prediction_data.get("remaining_life_hours", 0.0),
                    remaining_cycles=prediction_data.get("remaining_cycles", 0.0),
                    degradation_rate=prediction_data.get("degradation_rate", 0.0),
                    confidence_score=prediction_data.get("confidence_score", 0.5),
                    recommendations=prediction_data.get("recommendations", [])
                )
            else:
                logger.warning("Could not parse JSON from AI response")
                return self.predict_battery_life_analytical()
                
        except Exception as e:
            logger.error(f"Failed to parse AI response: {e}")
            return self.predict_battery_life_analytical()
    
    def calculate_degradation_rate(self, avg_health: float, avg_temp: float) -> float:
        """Calculate battery degradation rate"""
        # Base degradation rate
        base_rate = (100.0 - avg_health) / 100.0 * 0.1
        
        # Temperature factor
        temp_factor = 1.0
        if avg_temp > 45.0:
            temp_factor = 1.5 + (avg_temp - 45.0) * 0.1
        elif avg_temp < 10.0:
            temp_factor = 1.2
        
        return base_rate * temp_factor
    
    def calculate_remaining_life_hours(self) -> float:
        """Calculate remaining battery life in hours"""
        if len(self.performance_history) < 2:
            return 0.0
        
        # Calculate average current draw
        avg_current = sum(abs(d.pack_current) for d in self.performance_history) / len(self.performance_history)
        
        if avg_current <= 0:
            return 0.0
        
        # Estimate remaining capacity
        remaining_capacity = self.performance_history[-1].capacity_remaining / 100.0 * self.specs.nominal_capacity_ah
        
        return remaining_capacity / avg_current
    
    def calculate_average_health(self) -> float:
        """Calculate average battery health"""
        if not self.performance_history:
            return 0.0
        return sum(d.health_score for d in self.performance_history) / len(self.performance_history)
    
    def calculate_average_temperature(self) -> float:
        """Calculate average temperature"""
        if not self.performance_history:
            return 0.0
        return sum(d.cell_temp for d in self.performance_history) / len(self.performance_history)
    
    def calculate_temperature_range(self) -> float:
        """Calculate temperature range"""
        if not self.performance_history:
            return 0.0
        
        temps = [d.cell_temp for d in self.performance_history]
        return max(temps) - min(temps)
    
    def generate_recommendations(self, health: float, temp: float) -> List[str]:
        """Generate maintenance recommendations"""
        recommendations = []
        
        if health < 80:
            recommendations.append("Consider battery replacement soon")
        elif health < 90:
            recommendations.append("Monitor battery health closely")
        
        if temp > 45:
            recommendations.append("High temperature detected - check cooling system")
        elif temp < 10:
            recommendations.append("Low temperature - consider warming system")
        
        if len(self.performance_history) > 100:
            recommendations.append("Schedule preventive maintenance")
        
        return recommendations
    
    def enable_ai(self, api_key: str):
        """Enable AI predictions with API key"""
        self.gemini_api_key = api_key
        self.ai_enabled = True
        logger.info("AI prediction enabled")
    
    def disable_ai(self):
        """Disable AI predictions"""
        self.ai_enabled = False
        logger.info("AI prediction disabled")
    
    def get_prediction_report(self) -> Dict:
        """Get latest prediction report"""
        if not self.prediction_history:
            return {"error": "No predictions available"}
        
        latest = self.prediction_history[-1]
        return {
            "timestamp": latest.timestamp,
            "remaining_life_hours": round(latest.remaining_life_hours, 1),
            "remaining_cycles": round(latest.remaining_cycles, 0),
            "degradation_rate": round(latest.degradation_rate, 3),
            "confidence_score": round(latest.confidence_score, 2),
            "recommendations": latest.recommendations or [],
            "ai_enabled": self.ai_enabled
        }

# Example usage
if __name__ == "__main__":
    async def test_predictor():
        # Initialize predictor
        predictor = BatteryAIPredictor()
        
        # Add sample data
        for i in range(20):
            data = BatteryPerformance(
                timestamp=datetime.now().timestamp() + i * 3600,
                pack_voltage=350.0 + (i % 10) * 2.0,
                pack_current=50.0 + (i % 5) * 10.0,
                cell_temp=25.0 + (i % 3) * 5.0,
                capacity_remaining=85.0 - i * 0.5,
                cycle_count=i * 10,
                age_months=6.0 + i * 0.5,
                health_score=90.0 - i * 0.8
            )
            await predictor.add_performance_data(data)
        
        # Make prediction
        prediction = await predictor.predict_battery_life()
        
        # Print results
        report = predictor.get_prediction_report()
        print("=== AI Battery Life Prediction Report ===")
        print(json.dumps(report, indent=2))
        print("==========================================")
    
    # Run test
    asyncio.run(test_predictor()) 